src/main/java/com/metamx/cache/CaffeineCache.java

```diff
@@ -54,58 +55,79 @@ public static CaffeineCache create(final CaffeineCacheConfig config)
           .expireAfterAccess(config.getExpiration(), TimeUnit.MILLISECONDS);
     }
     if (config.getMaxSize() >= 0) {
-      builder = builder
-          .maximumSize(config.getMaxSize())
-          .maximumWeight(config.getMaxSize());
+      builder = builder.maximumWeight(config.getMaxSize());
+      builder.weigher(new Weigher<NamedKey, byte[]>()
+      {
+        @Override
+        public int weigh(@Nonnull NamedKey key, @Nonnull byte[] value)
+        {
+          return value.length;
+        }
+      });
+    }
+
+    final Cache delegateCache = config.getDelegateCache();
+    final CacheLoader<NamedKey, byte[]> cacheLoader;
+    if (delegateCache != null) {
+      cacheLoader = new CacheLoader<NamedKey, byte[]>()
+      {
+        @Override
+        public byte[] load(@Nonnull NamedKey key) throws Exception
+        {
+          return delegateCache.get(key);
+        }
+
+        @Override
+        public Map<NamedKey, byte[]> loadAll(@Nonnull Iterable<? extends NamedKey> keys) throws Exception
+        {
+          final Iterable<NamedKey> keys0 = (Iterable<NamedKey>) keys;
+          return Maps.transformValues(delegateCache.getBulk(keys0), CaffeineCache::serialize);
```

Comment:
You should probably wrap this in an `ImmutableMap.copyOf()` to ensure that serialization happens once per key. It would be easily error prone that the consumer would read the map multiple times and not expect the high penalty due to lazy transformations.

This is currently the case, so I might try to fix that to avoid your scenario from being a problem. Right now I do,
```java
Map<K, V> loaded = (Map<K, V>) cacheLoader().loadAll(keysToLoad);
loaded.forEach((key, value) -> {
  cache().put(key, value, false);
});
for (K key : keysToLoad) {
  V value = loaded.get(key);
  if (value != null) {
    result.put(key, value);
  }
}
```
That could probably be a single loop where a check to `keysToLoad.contains(key)` is used.